File: lab_02_25/src/c-func/scaler.c
#include <stdio.h>
#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE
float find_new_x(float kx, float now_x, float x, float width){
    return kx * (now_x - width / 2) + x * (1 - kx);
}


EMSCRIPTEN_KEEPALIVE
float find_new_y(float ky, float now_y, float y, float height){
    return ky * (height / 2 - now_y) + y * (1 - ky);
}

File: lab_02_25/src/scaler.js
import { Module } from './c-func/info.js';
import { StackPush } from './stack.js';

const width = window.innerWidth;
const height = window.innerHeight;

export function MakeScale(Group, layer, stack){
    var el = document.getElementById('scaling');
    var x = Number(el.x.value);
    var y = Number(el.y.value);
    var kx = Number(el.kx.value);
    var ky = Number(el.ky.value);

    var now_x = Group.getAttr('x');
    var now_y = Group.getAttr('y');

    var new_x = Module._find_new_x(kx, now_x, x, width); /// координаты нового центра в адекватных величинах
    var new_y = Module._find_new_y(ky, now_y, y, height);

    Group.setAttr('x', x + width / 2); //// ставим новый центр туда относительно которой точки масштабируем
    Group.setAttr('y', height / 2 - y);

    Group.setAttr('scaleX', kx * Group.getAttr('scaleX'));
    Group.setAttr('scaleY', ky * Group.getAttr('scaleY'));

    Group.setAttr('x', new_x + width / 2);
    Group.setAttr('y', height / 2 - new_y);

    layer.draw();
    StackPush(stack, Group);
}
